pyqt-apps/siriushla/widgets/spectrogram_view.py

```diff
@@ -0,0 +1,794 @@
+        if labels is None and (minVal is None or maxVal is None):
+            self._setLabels(self.labels)
+        elif labels is None:
+            self._setLabels({str(minVal): 0, str(maxVal): 1})
+        else:
+            self._setLabels(labels)
+
+    def _setGradient(self, g):
+        self.gradient = g
+        self.update()
+
+    def _setLabels(self, l):
+        """
+        Define labels to appear next to the color scale.
+
+        Accepts a dict of {text: value} pairs.
+        """
+        self.labels = l
+        self.update()
+
+    def paint(self, p, opt, widget):
+        GraphicsWidget.paint(self, p, opt, widget)
+        rect = self.boundingRect()  # Boundaries of visible area in scene coord
+        size = self.size if self.size is not None \
+            else [rect.right()-rect.left(), rect.bottom() - rect.top()]
+
+        # determine max width of all labels
+        textPadding = 2  # in px
+        labelWidth = 0
+        labelHeight = 0
+        for k in self.labels:
+            b = p.boundingRect(QRectF(0, 0, 0, 0),
+                               Qt.AlignLeft | Qt.AlignVCenter, str(k))
+            labelWidth = max(labelWidth, b.width())
+            labelHeight = max(labelHeight, b.height())
+
+        x1_back = (rect.right() - rect.left() - size[0])/2
+        x2_back = (rect.right() - rect.left() + size[0])/2
+        y1_back = (rect.bottom() - rect.top() - size[1])/2
+        y2_back = (rect.bottom() - rect.top() + size[1])/2
+
+        x1_grad = x1_back + textPadding*2
+        x2_grad = x2_back - (labelWidth + textPadding*2)
+        y1_grad = y1_back + (labelHeight/2 + textPadding)
+        y2_grad = y2_back - (labelHeight/2 + textPadding)
+
+        # Draw background
+        p.setPen(QPen(QColor(0, 0, 0, 0)))
+        p.setBrush(QBrush(QColor(255, 255, 255, 0)))
+        rect = QRectF(QPointF(x1_back, y1_back),
+                      QPointF(x2_back, y2_back))
+        p.drawRect(rect)
+
+        # Draw color bar
+        self.gradient.setStart(0, y2_grad)
+        self.gradient.setFinalStop(0, y1_grad)
+        p.setPen(QPen(QColor(0, 0, 0)))
+        p.setBrush(self.gradient)
+        rect = QRectF(QPointF(x1_grad, y1_grad),
+                      QPointF(x2_grad, y2_grad))
+        p.drawRect(rect)
+
+        # draw labels
+        tx = x2_grad + textPadding*2
+        lw = labelWidth
+        lh = labelHeight
+        for k in self.labels:
+            y = y1_grad + self.labels[k] * (y2_grad-y1_grad)
+            p.drawText(QRectF(tx, y - lh/2.0, lw, lh),
+                       Qt.AlignLeft | Qt.AlignVCenter, str(k))
+
+
+class SpectrogramUpdateThread(QThread):
+    """Thread to update image."""
+
+    updateSignal = pyqtSignal(list)
+
+    def __init__(self, spectrogram_view):
+        """Initialize Thread."""
+        QThread.__init__(self)
+        self.spectrogram_view = spectrogram_view
+
+    def run(self):
+        """Thread main."""
+        img = self.spectrogram_view.image_waveform
```

Comment:
If you want to be sure that this will run in a different thread, adding this line here:

`print('redraw ', int(QThread.currentThreadId()))`

and an equivalent `print` in the widget initialization and look if the integers printed are different.
I am asking this because I had problems with `QThread` recently. It was a different use case, however...

Reply:
I performed the test you suggested! The ''main'' and the ''redraw'' QThreads have different Ids, so it is in fact running in a separated thread :)